provider/cluster_rosa_classic_resource.go

```diff
@@ -1377,21 +1379,30 @@ func (r *ClusterRosaClassicResource) upgradeClusterIfNeeded(ctx context.Context,
 
 		clusterClient := r.clusterCollection.Cluster(state.ID.Value)
 		upgradePoliciesClient := clusterClient.UpgradePolicies()
-		gateIDs, description, err := upgrade.CheckMissingAgreements(desiredVersion.String(), state.ID.Value, upgradePoliciesClient)
-
-		if len(gateIDs) > 0 {
-			if !plan.UpgradeAdminAck.Unknown && !plan.UpgradeAdminAck.Null && plan.UpgradeAdminAck.Value {
-				for _, gateID := range gateIDs {
-					gateAgreementsClient := clusterClient.GateAgreements()
-					err := upgrade.AckVersionGate(gateAgreementsClient, gateID)
-					if err != nil {
-						return fmt.Errorf("failed to acknowledge version gate '%s' for cluster '%s': %v",
-							gateID, state.ID.Value, err)
-					}
-				}
-			} else {
-				return fmt.Errorf("%s\nTo acknoledge on this, please add \"upgrade_admin_ack = true\""+
-					" and re-apply the changes", description)
+		// Gates that require acknowledgement
+		ackGateIDs, description, err := upgrade.CheckMissingAgreements(desiredVersion.String(), state.ID.Value, upgradePoliciesClient, true)
+		if err != nil {
+			return fmt.Errorf("failed to check for missing upgrade agreements: %v", err)
+		}
+		targetMinorVersion := getOcmVersionMinor(desiredVersion.String())
+		userAcksGates := !plan.UpgradeAcksFor.Unknown && !plan.UpgradeAcksFor.Null && plan.UpgradeAcksFor.Value == targetMinorVersion
+		if len(ackGateIDs) > 0 && !userAcksGates { // User has not acknowledged mandatory gates
+			return fmt.Errorf("%s\nTo acknowledge these items, please add \"upgrade_acknowledgements_for = %s\""+
+				" and re-apply the changes", description, targetMinorVersion)
+		}
+
+		// Ack all gates to OCM; including ones that don't require _user_ acks
+		allGates, _, err := upgrade.CheckMissingAgreements(desiredVersion.String(), state.ID.Value, upgradePoliciesClient, false)
```

Comment:
Isn't it better to call this function only once and return all gates
In this case, the function can return in adition another bollean indicates if there are ackGates which require showing message to user, as it won't be possible to use `len(ackGateIDs)`  for that